src/backend/distributed/commands/function.c

```diff
@@ -187,22 +187,62 @@ create_distributed_function(PG_FUNCTION_ARGS)
 	EnsureFunctionOwner(funcOid);
 
 	ObjectAddressSet(functionAddress, ProcedureRelationId, funcOid);
-	ErrorIfFunctionDependsOnExtension(&functionAddress);
 
 	/*
-	 * when we allow propagation within a transaction block we should make sure to only
-	 * allow this in sequential mode
+	 * If the function is owned by an extension, only update the
+	 * pg_dist_object, and not propagate the CREATE FUNCTION. Function
+	 * will be created by the virtue of the extension creation.
 	 */
-	EnsureSequentialModeForFunctionDDL();
+	if (IsObjectAddressOwnedByExtension(&functionAddress, &extensionAddress))
+	{
+		if (IsObjectDistributed(&functionAddress))
+		{
+			ereport(ERROR, (errmsg("Function(%s) is already distributed.",
+								   get_func_name(funcOid))));
+		}
+
+		/*
+		 * Ensure corresponding extension is in pg_dist_object.
+		 */
+
+		Oid citusExtensionOid = get_extension_oid("citus", false);
+
+		if (citusExtensionOid == extensionAddress.objectId)
+		{
+			/* Citus extension is a special case */
+			ereport(ERROR, (errmsg("Citus extension functions(%s) "
+								   "can't be distributed.",
+								   get_func_name(funcOid))));
+		}
+		else if (!IsObjectDistributed(&extensionAddress))
+		{
+			ereport(WARNING, (errmsg("Extension(%s) owning the "
+									 "function(%s) is not distributed, "
+									 "attempting to propogate the extension",
+									 get_extension_name(extensionAddress.objectId),
+									 get_func_name(funcOid))));
+			EnsureDependenciesExistOnAllNodes(&functionAddress);
+		}
```

Comment:
This could/should be simplified to simply always `EnsureDependenciesExistOnAllNodes` like we do for many other operations.

If it has debug value to know we are propagating functions that could be embedded in that function so we have consistent reporting on it. Although I don't think we have found a need for that before

```suggestion
		EnsureDependenciesExistOnAllNodes(&functionAddress);
```

I don't know if it is possible to have a function defined by an extension rely on any object that is not covered by the creation of the extension. And while writing I think it will be impossible to hit a case where the distribution of an extension would not suffice to get all its dependencies distributed.

Simply calling Ensure is save and convenient. It reduces the mental overhead of thinking about edgecases. After this line has passed we know for sure anything required remotely is distributed. I guess if the extension is already distributed this call will always be a no-op.

Reply:
I've made the change